lib/api/core/validation/index.js

```diff
@@ -0,0 +1,788 @@
+    }
+
+    id = requestObject.data._id;
+  }
+
+  return this.validationPromise(index, collection, controller, action, id, requestObject.data.body, false)
+    .then(validationObject => {
+      if (!validationObject.validation) {
+        throw new BadRequestError(validationObject.errorMessages.join('\n'));
+      }
+
+      // We only modify the requestObject if the validation succeeds
+      requestObject.data.body = validationObject.documentBody;
+
+      return requestObject;
+    });
+};
+
+/**
+ * @param {string} index
+ * @param {string} collection
+ * @param {string} controller
+ * @param {string} action
+ * @param {string} id
+ * @param {*} documentCandidate
+ * @param {string[]|*} structuredError
+ * @returns {Promise.<{documentBody: *, errorMessages:string[], validation: boolean}>}
+ */
+Validation.prototype.validationPromise = function (index, collection, controller, action, id, documentCandidate, structuredError) {
+  var
+    // Avoid side effects on the requestObject during the validation
+    requestBodyClone = JSON.parse(JSON.stringify(documentCandidate)),
+    updateBodyClone;
+
+  structuredError = typeof structuredError !== 'undefined' && structuredError;
+
+  return new Promise((resolve, reject) => {
+
+    if (controller === 'write' && action === 'update') {
+      this.kuzzle.services.list.storageEngine.get({
+        index,
+        collection,
+        controller: 'read',
+        action: 'get',
+        data: {
+          _id: id
+        }
+      }).then(document => {
+        // Avoid side effects on the requestObject during the validation (on update)
+        updateBodyClone = JSON.parse(JSON.stringify(requestBodyClone));
+        _.defaultsDeep(requestBodyClone, document._source);
+        return resolve(requestBodyClone);
+      }).catch(error => {
+        return reject(error);
+      });
+    }
+    else {
+      return resolve(documentCandidate);
+    }
+  }).then(documentBody => {
+    var
+      validation = true,
+      strictnessError = false,
+      errorMessages = structuredError ? {} : [],
+      /** @type {CollectionSpecification} */
+      collectionSpec;
+
+    if (this.specification && this.specification[index] && this.specification[index][collection]) {
+      collectionSpec = this.specification[index][collection];
+
+      if (collectionSpec.fields) {
+        try {
+          validation = validation && this.recurseFieldValidation(documentBody, updateBodyClone || null, collectionSpec.fields.children, collectionSpec.strict, errorMessages, structuredError);
+        }
+        catch (error) {
+          if (error.message === 'strictness') {
+            // The strictness message can be received here only if it happens at the validation of the document's root
+            addErrorMessage('document', errorMessages, 'The document validation is strict; it can not add unspecified sub-fields.', structuredError);
+            strictnessError = true;
```

Comment:
I had 2 choices, stop at the first error, or report all errors. I chose to report all errors. And yes, there is a high change the DSL error can come from malformed or missing fields, but we can't tell ourselves so ...